linkerd/buffer/src/lib.rs

```diff
@@ -17,33 +18,40 @@ struct InFlight<Req, F> {
 pub(crate) fn new<Req, S>(
     inner: S,
     capacity: usize,
-) -> (Buffer<Req, S::Future>, Dispatch<S, Req, S::Future>)
+    idle_timeout: Option<Duration>,
+) -> (
+    Buffer<Req, S::Future>,
+    impl std::future::Future<Output = ()> + Send + 'static,
+)
 where
     Req: Send + 'static,
     S: tower::Service<Req> + Send + 'static,
-    S::Error: Into<Error>,
+    S::Error: Into<Error> + Send + 'static,
     S::Response: Send + 'static,
     S::Future: Send + 'static,
 {
+    use futures::future;
+
     let (tx, rx) = mpsc::channel(capacity);
     let (ready_tx, ready_rx) = watch::channel(Poll::Pending);
-    let dispatch = Dispatch::new(inner, rx, ready_tx);
+    let idle = move || match idle_timeout {
+        Some(t) => future::Either::Left(dispatch::idle(t)),
+        None => future::Either::Right(future::pending()),
+    };
```

Comment:
Any reason this is a closure and not just pass it in directly to `dispatch::run`? If passed directly it wouldn't need to to evaluate it each iteration

Reply:
Oh is this so `idle` can be repeatedly polled in the loop? This would move the value if we passed in `idle: I` instead of `idle: Fn() -> I`

Reply:
Right, this is a factory for idles, so they can be reset every time we start waiting for a request.